test/apps/test_hooks.c

```diff
@@ -88,88 +84,78 @@ static uint64_t get_time_msec(void)
 
 static void *thread_handler(void *arg)
 {
-    struct context *ctx = arg;
-    void *ptr;
-    uint64_t start, current;
+    context_t *ctx    = arg;
+    uint64_t deadline = get_time_msec() + ctx->thread_timeout_msec;
 
     /*
      * Many threads racing for read lock are preventing ucm to finish testing
      * the installed hooks, so we make sure to stop the threads after
      * reasonable time.
      */
-    start   = get_time_msec();
-    current = start;
-
-    while (!ctx->stop && ((ctx->thread_timeout_msec == 0) ||
-                          (current < (start + ctx->thread_timeout_msec)))) {
-        ptr = mmap_call(0);
-        if (ptr != (void*)-1) {
-            log("failed to return mmap_call() failure");
+    do {
+        if (mmap_call(0) != MAP_FAILED) {
+            printf("failed to return mmap_call() failure\n");
             exit(EXIT_FAILURE);
         }
-
-        current = get_time_msec();
-    }
+    } while (!ctx->stop &&
+             ((ctx->thread_timeout_msec == 0) || (get_time_msec() < deadline)));
 
     return NULL;
 }
 
-static ucs_status_t ctx_start_threads(struct context *ctx)
+static void ctx_stop_threads(context_t *ctx)
+{
+    int i;
+
+    printf("stopping %d threads\n", ctx->thread_count);
+
+    ctx->stop = 1;
+    for (i = 0; i < ctx->thread_count; i++) {
+        pthread_join(ctx->threads[i], NULL);
+    }
+}
+
+static ucs_status_t ctx_start_threads(context_t *ctx)
 {
     int i, ret;
 
     for (i = 0; i < ctx->thread_count; i++) {
-        ret = pthread_create(ctx->thread + i, NULL, thread_handler, ctx);
+        ret = pthread_create(&ctx->threads[i], NULL, thread_handler, ctx);
         if (ret < 0) {
-            log("pthread_create(thread #%d) failed: %m", i);
+            printf("pthread_create(thread #%d) failed: %m\n", i);
             goto err;
         }
     }
 
-    log("started %d threads", ctx->thread_count);
-
+    printf("started %d threads\n", ctx->thread_count);
     return UCS_OK;
 
 err:
-    for (; i > 0; i--) {
-        pthread_join(ctx->thread[i - 1], NULL);
-    }
-
+    ctx->thread_count = i;
```

Comment:
minor: IMO would be nicer to pass "count" parameter to ctx_stop_threads

Reply:
fixed